manager/allocator/network.go

```diff
@@ -386,13 +386,16 @@ func (a *Allocator) doNodeAlloc(ctx context.Context, nc *networkContext, ev even
 func serviceAllocationNeeded(s *api.Service, nc *networkContext) bool {
 	// Service needs allocation if:
 	// Spec has network attachments and endpoint resolution mode is VIP OR
-	// Spec has non-zero number of exposed ports and ingress routing is SwarmPort
+	// Spec has non-zero number of exposed ports and ingress routing is SwarmPort OR
+	// Spec has zero exposed ports and service has some
 	if (len(s.Spec.Networks) != 0 &&
 		(s.Spec.Endpoint == nil ||
 			(s.Spec.Endpoint != nil &&
 				s.Spec.Endpoint.Mode == api.ResolutionModeVirtualIP))) ||
 		(s.Spec.Endpoint != nil &&
-			len(s.Spec.Endpoint.Ports) != 0) {
+			len(s.Spec.Endpoint.Ports) != 0) ||
+		(s.Endpoint != nil && s.Spec.Endpoint != nil &&
+			len(s.Endpoint.Ports) != len(s.Spec.Endpoint.Ports)) {
```

Comment:
How about

```go
		(s.Endpoint != nil && len(s.Endpoint.Ports) != 0 &&
			(s.Spec.Endpoint == nil || len(s.Spec.Endpoint.Ports) == 0))
```

so that a nil `s.Spec.Endpoint` trigers reallocation when the the service has ports allocated.